planner/core/rule_partition_processor.go

```diff
@@ -106,8 +109,23 @@ type partitionTable interface {
 }
 
 func generateHashPartitionExpr(ctx sessionctx.Context, pi *model.PartitionInfo, columns []*expression.Column, names types.NameSlice) (expression.Expression, error) {
+	defer func() {
+		r := recover()
+		if r != nil {
+			logutil.BgLogger().Error("panic when generateHashPartitionExpr", zap.Reflect("r", r),
+				zap.Stack("stack trace"), zap.String("partExp", pi.Expr), zap.Reflect("pi", pi))
+			panic(fmt.Sprintf("%+v", r))
+		}
+	}()
 	schema := expression.NewSchema(columns...)
-	exprs, err := expression.ParseSimpleExprsWithNames(ctx, pi.Expr, schema, names)
+	colExprs, warns, err := expression.PartitionExtractExpr(pi)
+	if err != nil {
+		return nil, util.SyntaxWarn(err)
+	}
+	for _, warn := range warns {
+		ctx.GetSessionVars().StmtCtx.AppendWarning(util.SyntaxWarn(warn))
+	}
+	exprs, err := expression.RewriteExprsWithNames(ctx, colExprs, schema, names)
```

Comment:
The `RewriteExprsWithNames` is not thread-safe.
Rewrite an `ast.Ast` to `expression.Expression`, I thought it should be allocating new expression, thus thread-safe. But it turned out the `simpleRewriter` visit the at node is not thread-safe.

The old code chose to use `ParseSimpleExprsWithNames`, because from string to `expression.Expression` is thread-safe.

Reply:
Ref https://github.com/pingcap/tidb/pull/17353